lib/Basic/Unix/TaskQueue.inc

```diff
@@ -555,34 +562,36 @@ static bool
 cleanUpAHungUpTask(Task &T,
                    const TaskQueue::TaskFinishedCallback FinishedCallback,
                    const TaskQueue::TaskSignalledCallback SignalledCallback) {
-  const Optional<int> StatusIfOK = waitForPid(T.getPid());
-  if (!StatusIfOK)
+  const auto StatusAndProcessInformation = waitForPid(T.getPid());
+  if (!StatusAndProcessInformation.first)
     return true;
 
   T.finishExecution();
-  int Status = *StatusIfOK;
+  int Status = *(StatusAndProcessInformation.first);
+  TaskProcessInformation Process = StatusAndProcessInformation.second;
   return WIFEXITED(Status)
-             ? cleanUpAfterExit(Status, T, FinishedCallback)
+             ? cleanUpAfterExit(Status, T, Process, FinishedCallback)
              : WIFSIGNALED(Status)
-                   ? cleanUpAfterSignal(Status, T, SignalledCallback)
+                   ? cleanUpAfterSignal(Status, T, Process, SignalledCallback)
                    : false /* Can this case ever happen? */;
 }
 
-static Optional<int> waitForPid(const pid_t pidToWaitFor) {
+static std::pair<Optional<int>, TaskProcessInformation> waitForPid(const pid_t pidToWaitFor) {
   for (;;) {
     int Status = 0;
-    const pid_t pidFromWait = waitpid(pidToWaitFor, &Status, 0);
+    struct rusage Usage;
```

Comment:
The `wait4` call isn't POSIX, it's mostly-standard on anything BSD-compat but it should probably be conditionalized. I suspect it's safe to hang this on the `HAVE_GETRUSAGE` probe already done over in LLVM, but it's a bit indirect; maybe add a specific `check_symbol_exists(wait4, "sys/wait.h", HAVE_WAIT4)` probe to swift's `CMakeLists.txt`?

Reply:
I tried different configurations but it didn't worked out like I expected. I added `check_symbol_exists(wait4 "sys/wait.h" HAVE_WAIT4)` to the CMakeLists.txt and expected that I can check with `#if defined(HAVE_WAIT4)` in the code but it didn't work (it evaluated to false while wait4 should be there on my system.
Could you please give me a hint:

Reply:
Thanks for your help offline, I added the check-symbol to the cmakelist.